File: ruby/controller.rb
#!/usr/bin/env ruby

# Emotion Editor
# Copyright Mika Turkia 2006
# 25-31.7.2006


class ModelObject

	attr_reader :realobject, :expected_utility, :times_seen

	def initialize(controller, realobject, expected_utility)
		@controller = controller
		@realobject = realobject
		@expected_utility = expected_utility
		@times_seen = 1
	end

	def update_utility(event)
		@times_seen += 1
		@expected_utility = ((@expected_utility.to_f * (@times_seen - 1) + event.utility) / @times_seen)
	end

	def to_s
		eutil = sprintf("%4.2f", @expected_utility)
		" ( #{@realobject.class} #{@realobject.oid} util=#{eutil} seen=#{times_seen.to_s} #{@controller.get_attitude(self)})"
	end
end


class Controller

	def initialize(body)
		@known_objects = []
		@event_history = []
		@body = body
	end

	def process_event(e)
		@event_history.push(e)
		emotion = produce_affect(e)
		update_modelobject(e)
		emotion
	end

	def get_modelobject(realobject)
		@known_objects.each do |o| return o if o.realobject === realobject end
		nil
	end

	def body
		@body 
	end

	def update_modelobject(event)
		return if not event.target_object === self.body
		@known_objects.each do |o|
			if o.realobject === event.causing_object 
				o.update_utility(event)
				return
			end
		end
		@known_objects.push(ModelObject.new(self, event.causing_object, event.utility))
	end

	def alive?
		@body.alive?
	end

	def total_expected_utility
		sum = 0.0
		@known_objects.each do |o| sum += o.expected_utility end
		sum
	end

	def avg_expected_utility
                return 0.0 if @known_objects.size == 0
		total_expected_utility/@known_objects.size
	end

	def get_attitude(modelobject)
		if modelobject.expected_utility > 0.0 then future_emotion = "desire/hope"; past_emotion = "like/love"
		elsif modelobject.expected_utility < 0.0 then future_emotion = "disgust/fear"; past_emotion = "dislike/hate"
		else future_emotion = "neutral"; past_emotion = "neutral" end
		"attitude: #{past_emotion}, future expectation: #{future_emotion})"
	end

	def get_mood
		return "neutral" if not avg_expected_utility 
		if total_expected_utility > 0
			"good"
		elsif total_expected_utility < 0
			"bad"
		else
			"neutral"
		end
	end

	def to_s
		eutil = sprintf("%4.2f", total_expected_utility) 
		" #{@body.class} #{@body.oid} expects an utility of #{eutil} (mood=#{get_mood})" +
		" and has seen #{@known_objects.size} objects:\n#{@known_objects.join("\n")}"
	end

	def produce_affect(e)

		return nil if not e.causing_object or not e.target_object or not self.alive?

		if self.body === e.target_object
			# must check also if self === e.causing_object
			produce_targets_affects(e)
		elsif self.body === e.causing_object
			produce_causings_affect(e)
		else
			produce_outsiders_affect(e)
		end
	end

	def produce_targets_affects(e)

		modelofcausing = e.target_object.get_modelobject(e.causing_object)

		if not modelofcausing 
			# unexpected events
			if e.utility > 0.0
				emotion = "delightment/ilahtuminen" 
			elsif e.utility < 0.0
				emotion = "fright/saikahdys"
			else
				emotion = "surprise"
			end
		else 
			# expected events
			if modelofcausing.expected_utility >= 0.0
				if e.utility > modelofcausing.expected_utility
					emotion = "joy/happiness (ilo/onni)"
				elsif e.utility < modelofcausing.expected_utility
					emotion = "disappointment/pettymys"
				else emotion = "neutral" end
			elsif modelofcausing.expected_utility < 0.0
				if e.utility > modelofcausing.expected_utility
					emotion = "relief (helpotus)"
				elsif e.utility < modelofcausing.expected_utility
					emotion = "sadness/distress (suru)"
				else emotion = "neutral" end
			end

			# emotion of target object towards causing object
			if e.utility >= modelofcausing.expected_utility
				if self.body === e.causing_object then 
					emotion += " and pride (ylpeys)"
				else
					emotion += " and gratitude/admiration (kiitollisuus/ihailu) towards #{e.causing_object.oid}"
				end	
			elsif e.utility < modelofcausing.expected_utility
				if self.body === e.causing_object then 
					emotion += " and remorse (katumus)"
				else
					emotion += " and anger/reproach (viha/inho) towards #{e.causing_object.oid}"
				end
			end
		end
	end

	def produce_causings_affect(e)

		modeloftarget = self.get_modelobject(e.target_object)
		return nil if modeloftarget == nil

		if modeloftarget.expected_utility >= 0.0
				if e.utility > 0.0
					emotion = "happy for (myotailo) towards #{e.target_object.oid} and pride"
				elsif e.utility < 0.0
					emotion = "pity/compassion towards #{e.target_object.oid} and remorse, shame and anger towards self "
				else emotion = "neutral" end
		elsif modeloftarget.expected_utility < 0.0
				if e.utility > 0.0
					emotion = " envy towards #{e.target_object.oid} and remorse, shame and anger towards self"
				elsif e.utility < 0.0
					emotion = "gloating over/schadenfreude (vahingonilo) towards #{e.target_object.oid} and pride"
				else emotion = "neutral" end
		end
		emotion
	end

	def produce_outsiders_affect(e)

		modelofcausing = self.get_modelobject(e.causing_object)
		modeloftarget = self.get_modelobject(e.target_object)
		return nil if not modeloftarget or not modelofcausing
		return nil if not modelofcausing.expected_utility

		if modeloftarget.expected_utility >= 0.0
				if e.utility > 0.0
					emotion = "happy for (myotailo) towards #{e.target_object.oid}"
					emotion += " and gratitude (kiitollisuus) towards #{e.causing_object.oid}" if not e.causing_object === e.target_object
				elsif e.utility < 0.0
					emotion = "pity/compassion (saali) towards #{e.target_object.oid}" 
					emotion += " and anger/reproach (viha/inho) towards #{e.causing_object.oid}" if not e.causing_object === e.target_object
				else
					emotion = "neutral (neutraali) towards both"
				end
		elsif modeloftarget.expected_utility < 0.0
				if e.utility > 0.0
					emotion = "envy (kateus) towards #{e.target_object.oid}"
					emotion += " and anger/reproach (viha/inho) towards #{e.causing_object.oid}" if not e.causing_object === e.target_object
				elsif e.utility < 0.0
					emotion = "gloating over/schadenfreude (vahingonilo) towards #{e.target_object.oid}"
					emotion += " and gratitude (kiitollisuus) towards #{e.causing_object.oid}" if not e.causing_object === e.target_object
				else
					emotion = "neutral (neutraali) towards both"
				end
		end
		emotion
	end
end

File: javascript/controller.js
// Emotion Editor
// Copyright Mika Turkia 2006-2007
// 25-31.7.2006/24-25.1.2007/12.5.2007


function modelobject (controller, realobject, expected_utility) {
  this.controller = controller
  this.realobject = realobject
  this.expected_utility = expected_utility
  this.times_seen = 1
  this.update_utility = update_utility
  this.toString = print_modelobject
}

function update_utility(event) {
  this.times_seen += 1
  this.expected_utility = ((this.expected_utility * (this.times_seen - 1) + event.utility) / this.times_seen)
}

function print_model2 (realobject) {
  for (var i = 0; i < this.controller.known_objects.length; i++) {
    if (this.controller.known_objects[i].realobject === realobject) { return this.controller.known_objects[i].toString() }
  }
  return "unknown"
}

function print_modelobject () {
  return "utility: " + this.expected_utility.toFixed(2) + " seen: " + this.times_seen + "<br />"+ this.controller.get_past_attitude(this) + " (" + this.controller.get_future_attitude(this) + ")"
}

function controller (body) {
  this.known_objects = new Array()
  this.body = body
  this.process_event = c_process_event
  this.get_model = get_model
  this.print_model2 = print_model2
  this.update_modelobject = update_modelobject
  this.total_expected_utility = total_expected_utility
  this.avg_expected_utility = avg_expected_utility
  this.get_past_attitude = get_past_attitude
  this.get_future_attitude = get_future_attitude
  this.get_mood = get_mood
  this.toString = print_controller
  this.produce_affect = produce_affect
  this.produce_targets_affect = produce_targets_affect
  this.produce_causings_affect = produce_causings_affect
  this.produce_outsiders_affect = produce_outsiders_affect
}

function c_process_event(ev) {
  emotion = this.controller.produce_affect(ev)
  if (ev.target_object === this) { this.controller.update_modelobject(ev) }
  if (emotion == null) { emotion = "none" }
  return emotion
}

function get_model(realobject) {
  for (var i = 0; i < this.known_objects.length; i++) {
    if (this.known_objects[i].realobject === realobject) { return this.known_objects[i] }
  }
  return null
}

function update_modelobject(event) {
  for (var i = 0; i < this.known_objects.length; i++) {
    if (this.known_objects[i].realobject === event.causing_object) { 
      this.known_objects[i].update_utility(event)
      return null
    }
  }
  this.known_objects[this.known_objects.length] = new modelobject(this, event.causing_object, event.utility)
  this.known_objects.sort(sortModels)
  return null 
}

function sortModels(a, b) {
  return a.realobject.oid - b.realobject.oid
}

function total_expected_utility () {
  if (this.known_objects.length == 0) { return 0.0 }
  sum = 0.0
  for (var i = 0; i < this.known_objects.length; i++)
  {
    sum += this.known_objects[i].expected_utility
  }
  return sum
}

function avg_expected_utility () {
  return this.total_expected_utility() / this.known_objects.length
}

function get_past_attitude(modelobject) {
  if (modelobject.expected_utility > 0.0) { return "like/love" }
  else if (modelobject.expected_utility < 0.0) { return "dislike/hate" }
  else { return "neutral" }
}

function get_future_attitude(modelobject) {
  if (modelobject.expected_utility > 0.0) { return "desire/hope" }
  else if (modelobject.expected_utility < 0.0) { return "disgust/fear" }
  else { return "neutral" }
}

function get_mood () {
  if (this.total_expected_utility() == null) { return "neutral" }
  if (this.total_expected_utility() > 0) { return "good" }
  else if (this.total_expected_utility() < 0) { return "bad" }
  return "neutral"
}

function print_controller () {
  var temp = "Expected rewards: " + this.total_expected_utility().toFixed(2) + "<br />Mood: " + this.get_mood() + "<br />Self: "
  for (var i = 0; i < this.known_objects.length; i++) {
    if (this.known_objects[i].realobject === this.body) { return temp + this.known_objects[i].toString() }
  }
  return temp + "value unknown"
}

function produce_affect(ev) {
  // this = controller
  if (this.body === ev.target_object) { return this.produce_targets_affect(ev) }
  else if (this.body === ev.causing_object) { return this.produce_causings_affect(ev) }
  else { return this.produce_outsiders_affect(ev) }
  return null
}

function produce_targets_affect(e) {
  modelofcausing = ev.target_object.controller.get_model(ev.causing_object)
  var emotion = ""

  if (!modelofcausing)
  {
   // unexpected events
   if (ev.utility > 0.0) { emotion = "delightment" }
   else if (ev.utility < 0.0) { emotion = "fright" }
   else { emotion = "surprise" }
  }
  else 
  {
    // expected events
    if (modelofcausing.expected_utility >= 0.0) 
    {
      if (ev.utility >= modelofcausing.expected_utility) { emotion = "satisfaction/joy" }
      else if (ev.utility < modelofcausing.expected_utility) { emotion = "disappointment" }
      else {  emotion = "neutral" }
    }
    else if (modelofcausing.expected_utility < 0.0) 
    {
      if (ev.utility > modelofcausing.expected_utility) { emotion = "relief" }
      else if (ev.utility <= modelofcausing.expected_utility) { emotion = "fears-confirmed/sadness" }
      else { emotion = "neutral" }
    }

    // emotion of target object towards causing object
    if (ev.utility > modelofcausing.expected_utility) 
    {
      if (this.body === ev.causing_object) { emotion += " and pride" }
      else { emotion += " and gratitude towards " + ev.causing_object.oid }
    }	
    else if (ev.utility < modelofcausing.expected_utility)
    {
      if (this.body === ev.causing_object) { emotion += " and remorse" }
      else { emotion += " and anger towards " + ev.causing_object.oid }
    }
  }
  return emotion
}

function produce_causings_affect(e) {

  modeloftarget = this.get_model(ev.target_object)

  if (!modeloftarget) { return null }

  var emotion = ""

  if (modeloftarget.expected_utility >= 0.0)
  {
    if (ev.utility > 0.0)
    {
      emotion = "happy for " + ev.target_object.oid + " and pride"
    }
    else if (ev.utility < 0.0)
    {
      emotion = "pity/compassion towards " + ev.target_object.oid + ", remorse and anger towards self "
    }
    else { emotion = "neutral" }
  }
  else if (modeloftarget.expected_utility < 0.0)
  {
    if (ev.utility > 0.0)
    {
      emotion = " envy towards " + ev.target_object.oid + ", remorse and anger towards self"
    }
    else if (ev.utility < 0.0)
    {
      emotion = "gloating over/schadenfreude towards " + ev.target_object.oid + " and pride"
    }
    else { emotion = "neutral" }
  }
  return emotion
}

function produce_outsiders_affect(e) {
  modelofcausing = this.get_model(ev.causing_object)
  modeloftarget = this.get_model(ev.target_object)
  if (!modeloftarget) { return null } 

  var emotion = ""

  if (modeloftarget.expected_utility >= 0.0)
  {
    if (ev.utility > 0.0)
    {
      emotion = "happy for " + ev.target_object.oid
      if (ev.causing_object !== ev.target_object) { emotion += " and gratitude towards " + ev.causing_object.oid }
    }
    else if (ev.utility < 0.0)
    {
      emotion = "pity/compassion towards " + ev.target_object.oid
      if (ev.causing_object !== ev.target_object) { emotion += " and anger towards " + ev.causing_object.oid }
    }
    else
    {
      emotion = "neutral towards both"
    }
  }
  else if (modeloftarget.expected_utility < 0.0)
  {
    if (ev.utility > 0.0)
    {
      emotion = "envy towards " + ev.target_object.oid
      if (ev.causing_object !== ev.target_object) { emotion += " and anger towards " + ev.causing_object.oid }
    }
    else if (ev.utility < 0.0)
    {
      emotion = "gloating over/schadenfreude towards " + ev.target_object.oid
      if (ev.causing_object !== ev.target_object) { emotion += " and gratitude towards " + ev.causing_object.oid }
    }
    else
    {
      emotion = "neutral towards both"
    }
  }
  return emotion
}


